### PR TITLE
Switch to CMake

### PR DESCRIPTION
**Pull request by AlexanderLanin**
**Motivation:**
* CMake is a de-facto-standard
* Lines involved into building are reduced to ~10%
* Improved support for tooling due to direct support of/via cmake
* Even to non-cmake experts (like myself) the new 'code' looks way more readable than the old files.

**Based upon:**
* #528 by @cristianadam, however this in addition includes support for all travis jobs and completely removes the old way of building to prevent confusion and double maintenance.
* [C++ Weekly - Ep 208 - The Ultimate CMake / C++ Quick Start](https://www.youtube.com/watch?v=YbgH7yat-Jo) by @lefticus
* [Professional CMake: A Practical Guide](https://crascit.com/professional-cmake/)
* 0 cmake experience

**Out of scope:**
* lots of disabled compiler warnings came to light, they are not fixed
* potential switch to conan instead of downloading libb2 and zstd via cmake
* although interesting I'll put GitHub actions here, this is big enough as it is

**Currently missing:**
* lots of ToDos & polishing before review
* Anyone using .travis.sh, make travis or make docker in any way that should be supported here? Or are those options outdated anyway? I think this is the last thing missing from this PR.
**If needed** I would retry converting them again. Probably completely independent of make into a docker.sh/py. Otherwise I will remove them completely incl buildenv.
* **Which use case is missing? Please leave a comment.**

**Status:**
| Usecase | Old | Cmake |
| --------- | ---- | -------- |
| Build ccache | AC* && make -j ccache | cmake . && make ccache |
| Build and run tests | AC* && make -j test | cmake . && make check |
| Clang Tidy | AC* && make compile_commands.json && make tidy | cmake . -DENABLE_ CLANG_TIDY=ON && make |
| CppCheck | AC* && make cppcheck | cmake . -DENABLE_ CPPCHECK=ON && make |
| Clang Analyzer | AC* && make analyze | scan-build cmake . && scan-build make |
| Check formatting | AC* && make check_format | cmake . && make check_format |
| format | AC* && make format | cmake . && make format |
| ASAN tests | AC* && CFLAGS="-fsanitize=undefined" CXXFLAGS="-fsanitize=undefined" LDFLAGS="-fsanitize=undefined" make check | cmake . -DENABLE_SANITIZER_ UNDEFINED_BEHAVIOR=ON && make check |
| Compile and install | AC* && make install | cmake . && make install |
| Create source tgz | AC* && make dist | cmake . && make package_source |
| Create binary tgz | - | cmake . && make package |

Notes:
* AC*: ./autoconf && ./configure
* Those cmake ENABLE_...*: can be switched via GUI by running ccmake.
* Instead of cmake && make you can use cmake --build as it allows other defaults than make

**Travis:**
Added jobs:
* build source package, extract it, compile it, run tests
* build binary package, extract it, run tests
* update authors, build html docu

**Dependencies:**
Removed: compiledb
New: cmake, cmake-format, optionally ccmake

### CODE REVIEW

**Comment by cristianadam** [comment]
Installing is supported:
```cmake
install(TARGETS ccache DESTINATION .)
```
Installs the `ccache` binary.  

```
cmake --install <build_directory> [--prefix /path/to/install]
```
An example with CMake 3.15+. With previous versions you would have to set `CMAKE_INSTALL_PREFIX` and either call the `install` target, or execute the `install_cmake.cmake` script.

The current Windows ccache 3.7.8+ binaries contain also license and documentation files, but as `adoc` and not processed with a tool like `asciidoctor`.

**Comment by AlexanderLanin** [comment]
Is anyone actually using the docker approach? The travis image doesn't even build. I don't want to spend huge effort supporting something that no one needs.

    Executing: /tmp/tmp.dnIJEND2We/gpg.1.sh --fetch-keys
    https://apt.llvm.org/llvm-snapshot.gpg.key
    gpgkeys: protocol `https' not supported
    gpg: no handler for keyserver scheme `https'
    gpg: WARNING: unable to fetch URI https://apt.llvm.org/llvm-snapshot.gpg.key: keyserver error
    The command '/bin/sh -c printf "deb https://apt.llvm.org/xenial/ llvm-toolchain-xenial-7 main\ndeb-src https://apt.llvm.org/xenial/ llvm-toolchain-xenial-7 main\n" > /etc/apt/sources.list.d/llvm-toolchain.list && apt-key adv --fetch-keys https://apt.llvm.org/llvm-snapshot.gpg.key && apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 1E9377A2BA9EF27F' returned a non-zero code: 2

**Comment by cristianadam** [comment]
If you want to add a GitHub actions build, have a look at this [commit](https://github.com/cristianadam/ccache/commit/dbac3d7d2f7b4452dad9f67ac9a53ef7be62c11f
) from my fork.

For more information about it I blogged at: [Using GitHub Actions with C++ and CMake](https://cristianadam.eu/20191222/using-github-actions-with-c-plus-plus-and-cmake/).

File: dev.mk.in [hunk 1]
```diff
@@ -1,245 +0,0 @@
...
-%.xml: %.adoc
-	@mkdir -p $(@D)
-# Make literals stand out as bold in the man page:
-	$(if $(quiet),@echo "  ASCIIDOC $@")
-	$(Q)$(ASCIIDOC) -a revnumber=$(version) -d manpage -b docbook -o - $< | \
-	  perl -pe 's!<literal>(.*?)</literal>!<emphasis role="strong">\1</emphasis>!g' >$@
-
-doc/ccache.1: doc/MANUAL.xml
-	$(if $(quiet),@echo "  A2X      $@")
-	$(Q)$(A2X) --doctype manpage --format manpage $<
```
Comment (AlexanderLanin) [inline dev.mk.in:177]: Could anyone explain what is happening here?
I don't understand what is actually executed between all this make magic.

**Comment by AlexanderLanin** [comment]
> Is anyone actually using the docker approach? The travis image doesn't even build. I don't want to spend huge effort supporting something that no one needs.
> 
> ```
> Executing: /tmp/tmp.dnIJEND2We/gpg.1.sh --fetch-keys
> https://apt.llvm.org/llvm-snapshot.gpg.key
> gpgkeys: protocol `https' not supported
> gpg: no handler for keyserver scheme `https'
> gpg: WARNING: unable to fetch URI https://apt.llvm.org/llvm-snapshot.gpg.key: keyserver error
> The command '/bin/sh -c printf "deb https://apt.llvm.org/xenial/ llvm-toolchain-xenial-7 main\ndeb-src https://apt.llvm.org/xenial/ llvm-toolchain-xenial-7 main\n" > /etc/apt/sources.list.d/llvm-toolchain.list && apt-key adv --fetch-keys https://apt.llvm.org/llvm-snapshot.gpg.key && apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 1E9377A2BA9EF27F' returned a non-zero code: 2
> ```

@jrosdahl are you using .travis.sh, make travis or make docker in any way that should be supported here? Or are those options outdated anyway? I think this is the last thing missing from this PR.
If needed I would retry converting them again. Probably completely independent of make into a docker.sh/py.

**Comment by AlexanderLanin** [comment]
I've got docker stuff working but noticed there is no c++ compiler included so far. So it's obviously totally outdated.

@jrosdahl: should i include docker + fixes here in this PR or remove it completely since it's unneeded complexity?

File: dev.mk.in (hunk 1 above)
Comment (jrosdahl) [reply dev.mk.in:177]: Are you talking about the [pattern rules](https://www.gnu.org/software/make/manual/html_node/Pattern-Rules.html) syntax/semantics or the commands they execute (the "recipes" in Make terminology)? What do you think is magic about them?

**Comment by jrosdahl** [comment]
Thanks for continuing the work.

> Lines involved into building are reduced to ~10%

Parenthetically, but still: I don’t think that’s a fair comparison. :slightly_smiling_face: `config.{guess,sub}`, `install-sh` and all files in `m4` are essentially third party code which you seldom need to care about.

The Autotools files you do need to care about (`autogen.sh`, `configure.ac`, `dev.mk.in` and `Makefile.in`) weigh in at ≈783 lines while the `CMakeLists.txt` and custom `*.cmake` files in the PR weigh in at ≈894 lines. It’s of course not exactly an apples to apples comparison, I just wanted to point out that size of the build system “configuration” is roughly the same, and I don’t think that it’s a relevant factor.

> Even to non-cmake experts (like myself) the new 'code' looks way more readable than the old files.

I also think that people with neither CMake nor Autotools (+ Makefile and shell scripting) experience will pick up CMake more easily.

Switching to CMake will likely optimize for new contributors but not for _me_, at least not in the short term since I don’t have much CMake experience. We also optimize for _some_ users but at the same time make things worse for other users. I just want to point out that this is a trade-off.

That said, I don’t particularly like Autotools and I think that CMake seems less bad than Autotools in several aspects.

> I've got docker stuff working but noticed there is no c++ compiler included so far. So it's obviously totally outdated.

The Docker build functionality was contributed by @afbjorklund in #167, #245 and #278. I seldom use it myself so I’m not married to it. As you have noticed one has to put a bit of effort into making it work again from time to time.

> @jrosdahl: should i include docker + fixes here in this PR or remove it completely since it's unneeded complexity?

I do like the original idea, but if we don’t have time to make it work I’m OK with removing it, at least for now. On the other hand you seem to have it kind of working already?

> Which use case is missing? Please leave a comment.

There is one important use case I miss:

| Use case | Old | CMake |
| --------- | ---- | -------- |
| Build ccache as an end user | ./configure && make && make install | cmake . && make |

By “end user” I here mean a user building from the source code release, i.e. a “non-dev build” in the current ccache build system. I would like the source code release archive to contain pregenerated documentation (man page and HTML files) just like today so that the end user doesn’t have to install asciidoc and xsltproc to get generated documentation. Would that be reasonable? Or is this something you never do in the CMake world? Also, `WARNINGS_AS_ERRORS` needs to default `FALSE` for end user builds. The set of extra warning flags sent to GCC maybe also needs to be trimmed so that older GCC versions don’t choke on them?

Give me a shout when you want me to review things in more detail or if you want me 

Here are some things I have noticed so far:

* `ccache --version` doesn’t print the version. That’s because `CCACHE_VERSION` has an incorrect type in `src/Version.cpp.in` (it should be `const char CCACHE_VERSION[]`, not `const char* CCACHE_VERSION`).
* When the test suite fails you don’t get the failure output by default, which is unfortunate. You can use `ctest --output-on-failure` but I would like that behavior to be the default, at least for `make test`. Would that be possible?
* I like the setup with one `addtest` per test suite. Nice!
* The generated version should be `<branch>.<hash>` for a development build, see e2cb5f8f3fdd183872703dbdbf8be63f5e02c7d9.
* The required zstd version is too strict; ccache supports older zstd versions as well. It seems that at least 1.1.2 is supported so I suggest using that as the lower bar.
* Similarly, I suggest 0.98 as the lower bar for libb2.
* Would it be OK or bad practise to make `CMAKE_BUILD_TYPE` default to `Release` or `RelWithDebInfo` (perhaps if `CFLAGS` and `CXXFLAGS` have not been set, as mentioned by @cristianadam in a comment to [_CMake and the Default Build Type_](https://blog.kitware.com/cmake-and-the-default-build-type/))? Not sure if it’s unfounded, but I’m a bit worried that an end user may just do “cmake . && make” without understanding that it will be a non-optimized build by default.
* Speaking of which, would it make sense to use release builds in Travis? I think so, but without `-DNDEBUG`.
* Can we have `cmake . && make test` work, i.e. have the test target depend on `ccache` and `unittest/run`?
* Don’t know if it’s work in progress, but I would like the source release archives to be named like before with the same layout as before. And, as mentioned above, to include generated documentation. And I would like a tar.xz source release archive as well. Oh, and a pony.
* We need to remember to 

FYI regarding releases: I have a number of local scripts for preparing, creating and signing a release, performing additional sanity checks, creating mailing list announcement, creating a GitHub release, etc. The experimental Windows binary packages are currently created ad hoc by those scripts. I'm OK with that so you don't have to put effort into creating binary releases.

**Comment by afbjorklund** [comment]
> > I've got docker stuff working but noticed there is no c++ compiler included so far. So it's obviously totally outdated.
> 
> The Docker build functionality was contributed by @afbjorklund in #167, #245 and #278. I seldom use it myself so I’m not married to it. As you have noticed one has to put a bit of effort into making it work again from time to time.

Feel free to delete the Dockerfiles, the Vagrantfiles, and the Travis emulator:

```
buildenv/alpine/Dockerfile
buildenv/centos/Dockerfile
buildenv/debian/Dockerfile
buildenv/fedora/Dockerfile
buildenv/ubuntu/Dockerfile
misc/freebsd/Vagrantfile
misc/build.sh
.travis.sh
.travis/Dockerfile
```

As well as the "docker" and "travis" build targets:
```Makefile
BUILDENV = ubuntu
DOCKER_IMAGE_TAG = ccache/build:$(BUILDENV)

.PHONY: docker
docker: buildenv/$(BUILDENV)/Dockerfile
	$(DOCKER) inspect $(DOCKER_IMAGE_TAG) >/dev/null || $(DOCKER) build -t $(DOCKER_IMAGE_TAG) buildenv/$(BUILDENV)
	$(DOCKER) run --rm -v $(PWD):/build -w /build $(DOCKER_IMAGE_TAG) misc/build.sh $(TEST)

.PHONY: travis
travis: .travis/Dockerfile
	$(DOCKER) inspect travis-build >/dev/null || $(DOCKER) build -t travis-build .travis
	$(DOCKER) run --rm --volume $(PWD):/src --tmpfs /dst:rw,exec --env ASAN_OPTIONS='$(ASAN_OPTIONS)' travis-build \
	sh -c "cd /src && ./autogen.sh && cd /dst && CC=$(CC) CFLAGS='$(CFLAGS)' CXX=$(CXX) CXXFLAGS='$(CXXFLAGS)' LDFLAGS='$(LDFLAGS)' /src/configure $(CONFIGURE) && CC=$(CC) CFLAGS='$(CFLAGS)' make V=$(V) && CC=$(CC) make V=$(V) $(TEST)"
```

Or at least update or recreate them, for the "new ccache" (i.e. the C++ version) ?

Also: https://hub.docker.com/r/ccache/build

I will **not** be updating them, so if nobody elses uses them you can clean them out.

If you keep them, probably settle on _one_...

**Comment by AlexanderLanin** [comment]
Thanks!

Docker turned out to be really surprisingly helpful! Never used it before. I've fixed some problems with different gcc and clang versions (could have done that without docker). But more interestingly tests fail under debian-9 gcc build. No idea whether that's somehow docker related yet. Anyway I'll move docker support to a separate PR as this one is getting just too big and it would delay me another few weeks.

Regarding the other feedback:
* I must admit cmake besides basic stuff is starting to get slightly tricky. Nevertheless I believe it to be a de-facto standard these days.

* > Switching to CMake will likely optimize for new contributors but not for me, at least not in the short term since I don’t have much CMake experience. We also optimize for some users but at the same time make things worse for other users. I just want to point out that this is a trade-off.

  That's why I started the table above to enable some sort of transition. It's not 1 to 1, so some help is needed for those that are switching. Maybe that should go to some readme file.

* Build ccache as an end user is actually tested on travis now as part of build_and_verify_package_source.sh. It generates the package, extracts it and compiles the source files.

I'll address the TODO list within the next days.

**Comment by AlexanderLanin** [comment]
I just spent three days of free time for rebasing/merging. It's clear I will not be able to keep this up.
I'll look into finalizing this SOON or splitting parts of which can be merged right away.

The TODOs are partially addressed (make check instead of test)

**Comment by jrosdahl** [comment]
> I just spent three days of free time for rebasing/merging. It's clear I will not be able to keep this up.
I'll look into finalizing this SOON or splitting parts of which can be merged right away.

Sorry to hear that it's hard to keep up (which is what I also feel from time to time, of course). Can I do anything to help? Work on rebasing the branch?

**Comment by AlexanderLanin** [comment]
> > I just spent three days of free time for rebasing/merging. It's clear I will not be able to keep this up.
> > I'll look into finalizing this SOON or splitting parts of which can be merged right away.
> 
> Sorry to hear that it's hard to keep up (which is what I also feel from time to time, of course). Can I do anything to help? Work on rebasing the branch?

Thanks, my time was/is somewhat limited at the moment. I've a bunch of uncommitted stuff which I will push here once I get it through travis. (Hopefully in the next minutes).
It would be nice to avoid travis changes until this is merged or abandoned.

**Comment by AlexanderLanin** [comment]
I think this might be "ready".
I'm sure it's not perfect. It can be improved afterwards!
But is it good enough for an official switch?

Experimental and not up to detailed review:
* Building binary package via make package
* Building source package via make package_source
* Docker support
(It's better/more than before, but not really done)

I know this is a huge change and if anyone sees an incremental approach I'm all for it!

Does anyone have an idea how to make check_format more verbose? Something like run-clang-format? It works fine on my PC, so I'm guessing Travis has a different version of some tool.

@cristianadam @agouaillard maybe you two could also have a look at this PR?

File: .travis.yml [hunk 2]
```diff
@@ -114,47 +108,65 @@ matrix:
...
+    env: T="Build binary and verify" SPECIAL=build_and_verify_package
     addons:
       apt:
         packages:
         - elfutils
         - libzstd1-dev
         - libb2-dev
 
-  # New jobs go here until they are established for some weeks
-  allow_failures:
```
Comment (jrosdahl) [inline .travis.yml:138]: Is it intentional to add Clang-Tidy and Clang-Format as required tests? I'm OK with that, but it feels like that should be a separate PR.

File: .travis.yml [hunk 3]
```diff
@@ -1,20 +1,15 @@
...
-  - make ${TEST:-test}
-
-matrix:
+script: ci/build.sh
+
+jobs:
   include:
 
-  # Job 1: Build (in separate build directory) on Linux with native GCC
+  # Job 1: Build on Linux with native GCC
```
Comment (jrosdahl) [inline .travis.yml:9]: Since CMake supports in-place builds (`cmake . && make ...`) I think that it would make sense to let at least one of the jobs test that it works since some end users surely will use that method.

File: doc/DEVELOPER.md [hunk 4]
```diff
@@ -7,7 +7,7 @@ Tracing
 In order to see what ccache is doing, it is possible to enable microsecond
 tracing:
 
-* Build ccache with the `--enable-tracing` configure option.
+* Build ccache with the `-DENABLE_TRACING` cmake option.
```
Comment (jrosdahl) [inline doc/DEVELOPER.md:10]:
```suggestion
* Build ccache with the `-DENABLE_TRACING=1` cmake option.
```

File: .travis.yml [hunk 5]
```diff
@@ -114,47 +108,65 @@ matrix:
...
-    env: T="Clang-Tidy" TEST=tidy V=1 PATH="/usr/bin:$PATH" NO_COMPILE=1
+    compiler: clang
+    env: T="Clang-Tidy" CMAKE_PARAMS="-DENABLE_CLANG_TIDY=ON" RUN_TESTS=0
+    addons:
+      apt:
+        packages:
+        - libzstd1-dev
+        - libb2-dev
+
+  # Job 12: Run Clang-Format amd CMake-Format check mode
```
[the remark this answers is not in the document]
Comment (AlexanderLanin) [reply .travis.yml:119]: Yeah I had the same trouble as travis was even more up to date than locally. You do get errors showing the diff, but it's still annoying to deal with. It seems to be more volatile than clang-format, however clang-format is also already starting to get annoying.
Maybe we should just disable the travis test for formatting completely and run format manually from time to time? I don't know.

File: cmake/config.h.in [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+#ifdef __clang__
+#  pragma clang diagnostic push
+#  if __has_warning("-Wreserved-id-macro")
+#    pragma clang diagnostic ignored "-Wreserved-id-macro"
+#  endif
+#endif
+
+#cmakedefine01 _GNU_SOURCE
+#cmakedefine01 _POSIX_C_SOURCE
+#cmakedefine01 _DARWIN_C_SOURCE
```
[the remark this answers is not in the document]
Comment (AlexanderLanin) [reply cmake/config.h.in:11]: Yes, indeed. It's really difficult to get them right without all those test environments.

File: misc/format.sh [hunk 7]
```diff
@@ -1,2 +1,8 @@
 #!/bin/sh -ex
 find src unittest -path src/third_party -prune -o -regex ".*\.[ch]p?p?" -exec misc/run-clang-format {} \;
+
+# Top level CMakeLists.txt + subidrectories.
+# This avoids running the check on any build directories.
+cmake-format -i CMakeLists.txt
```
[the remark this answers is not in the document]
Comment (AlexanderLanin) [reply misc/format.sh:6]: it should probably be if(cmake-format available) { format... }

File: cmake/CCachePackConfig.cmake [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+list(APPEND CPACK_SOURCE_IGNORE_FILES "^${CMAKE_SOURCE_DIR}/build[\\-_/]")
+list(APPEND CPACK_SOURCE_IGNORE_FILES "^${CMAKE_BINARY_DIR}")
+
+# A top level directory is nice for extracting in-place, but prevents extracting
+# directly to e.g. /usr/local. Unfortunately this is not split between binary
+# and source build, so there is always no top level directory.
+set(CPACK_INCLUDE_TOPLEVEL_DIRECTORY OFF)
+
+set(CPACK_PACKAGE_FILE_NAME "ccache-binary")
+set(CPACK_SOURCE_PACKAGE_FILE_NAME "ccache-src")
```
[the remark this answers is not in the document]
Comment (AlexanderLanin) [reply cmake/CCachePackConfig.cmake:44]: As mentioned I stopped improving the source and binary builds via cpack to prioritize on the other stuff first. However I did not want to remove what is working so far.

File: cmake/config.h.in (hunk 6 above)
[the remark this answers is not in the document]
Comment (AlexanderLanin) [reply cmake/config.h.in:11]: Hmm... https://cirrus-ci.org/guide/FreeBSD/
But then this PR will never end.

**Comment by AlexanderLanin** [comment]
> Yeah, we need to get this landed. It has of course taken much time for you, but also for me and I want to be able focus on "real" things for a while now. It feels like I have made very little progress on my goals for ccache 4.0 last 6 months due to various refactoring taking up most of my bandwidth.

Switching to more C++ style code is no small task. And it's risky one, so it must be done carefully. No new features, lots of potential bugs. All for a potential benefit in the future. Same as CMake...
If I had know how big this would get I would not have touched it.

> Also, I really hope that this is worth the effort in the long run. To me, many things seem uglier and more complex now...

I think that's mostly because it's cleaner - in a sense. For example compiler settings are now target specific. Which will help. But it's more difficult at the moment, compared to just setting them globally as before.

**Comment by jrosdahl** [comment]
@AlexanderLanin: What's your view on the current status of this PR?

**Comment by AlexanderLanin** [comment]
TODO in config.h.in is still missing about feature macros, I've addressed the other findings.

**Comment by jrosdahl** [comment]
> The feature macros are not 100% as they were, but I feel it's a waste of time.

Not sure I understand. Do you feel that it's a waste of time to try to support less popular environments or to try to reproduce the old set of feature test macros to 100%? I agree with the latter but not with the former if it's easy enough to do it.

> The far better approach is to remove all non C/C++/Posix function calls.

Which calls are you referring to?

> I've addressed all other findings.

OK, thanks. I'll look into merging it when I have some more time and then I hope we can collaborate on improving it. "Fortunately" a 4.0 release is still quite far away so we have time to fix missing parts like packaging of release archive.

**Comment by AlexanderLanin** [comment]
Those feature macros which have quite some logic depending on OS and OS-Version are only required due to non standard stuff like localtime_r.
localtime_r (and similar) are a good idea in general, but they require a bunch of different defines depending on OS & Version.

I think those functions should be removed one by one. In the localtime example ccache will not work any worse if time is off my a second due to some threading issues. So we can just replace localtime_r with localtime.


So short term having the feature macros not 100% the same as before might loose some support for some environments. But instead of getting those defines right, it seems better to not require them at all and use only standard C++11.

**Comment by jrosdahl** [comment]
> I think those functions should be removed one by one.

To "remove all non C/C++/Posix function calls" means removing Windows support and also things like file clone support, increased time resolution of stats in #577, progress bar width calculation, etc. Is that what you're proposing?

> In the localtime example ccache will not work any worse if time is off my a second due to some threading issues. So we can just replace localtime_r with localtime.

Regarding the specific case of `localtime_r`: the reason I started it instead of `localtime` had nothing to do with threading. It was to [please linters](https://github.com/ccache/ccache/commit/753ca2b9341b39d81e474065b92a958ad2aca021), in this case [LGTM](https://lgtm.com/projects/g/ccache/ccache/alerts). It would certainly be possible to change back to `localtime` and add an ignore instead if using `localtime_r` becomes a problem.

> So short term having the feature macros not 100% the same as before might loose some support for some environments.

That's OK. People interested in portability to more esoteric platforms will come by and suggest fixes.

> [...] only standard C++11.

That's impossible.

**Comment by AlexanderLanin** [comment]
Sorry I expressed myself poorly. As you wrote it’s totally fine and even necessary to have platform dependent code.
However I personally think it should be encapsulated to specific platforms and not be the case for the actual „core“. In this specific case it does get quite complex when you start needing a few dozens line of logic which #define needs to be set to include the proper part of some library. That is *probably* just too exotic and not supported widely enough. 
I have not yet looked at which functions exactly are affected. 
I‘m guessing some can be replaced easily and some may be almost impossible.

**Comment by jrosdahl** [comment]
I still don't understand what point you want to make.

**Comment by AlexanderLanin** [comment]
There is a reproducible bug which I do not know how to analyze by running: `./docker.sh debian-9-stretch`

    The following tests FAILED:
          2 - producttest.base (Failed)
          3 - producttest.nocpp2 (Failed)

**Comment by jrosdahl** [comment]
Yes, merging this has been on my to do list for a couple of weeks. I hope to do it soon.

**Comment by jrosdahl** [comment]
Thanks @AlexanderLanin and @cristianadam. I'll continue working on some aspects of it on master.
